cms/djangoapps/contentstore/models.py

```diff
@@ -115,31 +103,107 @@ class PublishableEntityLink(models.Model):
     created = manual_date_time_field()
     updated = manual_date_time_field()
 
-    def __str__(self):
-        return f"{self.upstream_usage_key}->{self.downstream_usage_key}"
+    class Meta:
+        abstract = True
 
     @property
     def upstream_version(self) -> int | None:
         """
         Returns upstream block version number if available.
         """
-        version_num = None
-        if hasattr(self.upstream_block, 'published'):
-            if hasattr(self.upstream_block.published, 'version'):
-                if hasattr(self.upstream_block.published.version, 'version_num'):
-                    version_num = self.upstream_block.published.version.version_num
-        return version_num
+        published_version = get_published_version(self.upstream_block.publishable_entity.id)
+        return published_version.version_num if published_version else None
 
     @property
     def upstream_context_title(self) -> str:
         """
         Returns upstream context title.
         """
-        return self.upstream_block.learning_package.title
+        return self.upstream_block.publishable_entity.learning_package.title
+
+    @classmethod
+    def filter_links(
```

Comment:
Non-blocking: This is something that seems like it should go into a custom Manager instead of a classmethod on the base class.